src/Components/Components/src/RenderTree/RenderTreeBuilder.cs

```diff
@@ -454,13 +454,30 @@ public void OpenComponent(int sequence, Type componentType)
         /// <param name="value">The value for the key.</param>
         public void SetKey(object value)
         {
-            // This is just a placeholder to enable work in parallel in the
-            // aspnetcore-tooling repo.
-            //
-            // The real implementation will involve multiple overloads, likely:
-            //   SetKey(int value) -- underlying logic
-            //   SetKey<T>(T value) where T: struct -- avoids boxing 'value' before calling .GetHashCode()
-            //   SetKey(object value) -- performs null check before calling .GetHashCode()
+            if (value == null)
+            {
+                throw new ArgumentNullException(nameof(value));
+            }
+
+            var parentFrameIndex = GetCurrentParentFrameIndex();
+            if (!parentFrameIndex.HasValue)
+            {
+                throw new InvalidOperationException("Cannot set a key outside the scope of a component or element.");
+            }
+
+            var parentFrameIndexValue = parentFrameIndex.Value;
+            ref var parentFrame = ref _entries.Buffer[parentFrameIndexValue];
+            switch (parentFrame.FrameType)
+            {
+                case RenderTreeFrameType.Element:
+                    _entries.Buffer[parentFrameIndexValue] = parentFrame.WithElementKey(value);
```

Comment:
out of curiousity would `parentFrame = parentFrame.WithElementKey(value);` work here?

Reply:
Yes it would, and I suppose theoretically it could be faster because it doesn't have to recompute the pointer to the array entry. Not sure it would really make any difference in practice.

Anyway I've changed it to work as you suggest, with a clarifying comment.